doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+* [`save_input_as`](#save_input_asvariable_name)
+* [`calculate`](#calculatevariable_name-block)
+
+Each of these block types and the point at which they are executed is explained in more detail below:
+
+##### `precalculate(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed to render a question or outcome node template.
+* It's simplest to think about these blocks executing before the user response is parsed from the request path.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> Note that due to an oversight in the implementation of `Flow#process`, it's not currently possible to use a `precalculate` block in the first question.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+##### `on_response(&block)`
+
+* These blocks are intended to be used to store user responses on a `calculator` state variable. They are a [relatively new addition to the DSL][introduction-of-on-response-blocks].
+* These blocks are called  after the user response has been parsed from the request path and before any of the `next_node_calculation` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is not used and no state variable is stored.
+
+> The use of these blocks is encouraged; however, they should only ever be used to store a single, `calculator` state variable (in the first question definition); otherwise they should only be used to store user responses on that `calculator` object.
+
+##### `next_node_calculation(variable_name, &block)`
+
+* These blocks were intended to be used to store state variables that are needed in the `next_node` block.
+* These blocks are executed after all the `on_response` blocks have been executed and before any of the `validate` blocks are executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* The block return value is stored on the state object as a state variable named `variable_name`.
+
+> The use of these blocks is deprecated and should never be necessary. Define methods on a `calculator` object instead.
+
+##### `validate(message_key, &block)`
+
+* These blocks are intended to be used to validate the user response.
+* These blocks are executed after all the `next_node_calculation` blocks have been executed and before the `next_node` block is executed.
+* The parsed response is passed to the block as the only argument and by convention is named `response`.
+* If the block return value is truth-y, then no action is taken.
+* If the block return value is false-y, then:
+  * A `SmartAnswer::InvalidResponse` exception is raised with the `message_key` set as the exception message.
+  * Although this exception is handled within the app, it prevents the transition to the next node.
```

Comment:
> This is part of the previous sub-bullet, and should be moved up.

I'm not sure this is right. The sub-bullets are intended to explain the rough sequence of events. I've tweaked the text and numbered the bullet points to try to make this clearer. See this fix-up commit: 4b087edc9a3559ba23037a4585b4c139a4a3229b.